packages/beacon-node/src/chain/blocks/blockInput/blockInput.ts

```diff
@@ -0,0 +1,786 @@
+      blockRootHex: props.blockRootHex,
+      parentRootHex: toHex(props.blobSidecar.signedBlockHeader.message.parentRoot),
+      slot: props.blobSidecar.signedBlockHeader.message.slot,
+    };
+    const blockInput = new BlockInputBlobs(init, state);
+    blockInput.blobsCache.set(props.blobSidecar.index, {
+      blobSidecar: props.blobSidecar,
+      source: props.source,
+      seenTimestampSec: props.seenTimestampSec,
+      peerIdStr: props.peerIdStr,
+    });
+    return blockInput;
+  }
+
+  getLogMeta(): LogMetaBlobs {
+    return {
+      blockRoot: prettyBytes(this.blockRootHex),
+      slot: this.slot,
+      expectedBlobs: this.state.hasBlock ? this.state.block.message.body.blobKzgCommitments.length : "unknown",
+      receivedBlobs: this.blobsCache.size,
+    };
+  }
+
+  addBlock({blockRootHex, block, source}: AddBlock<ForkBlobsDA>): void {
+    if (!this.state.hasBlock) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INVALID_CONSTRUCTION,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot addBlock to BlockInputBlobs after it already has a block"
+      );
+    }
+
+    // this check suffices for checking slot, parentRoot, and forkName
+    if (blockRootHex !== this.blockRootHex) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.MISMATCHED_ROOT_HEX,
+          blockInputRoot: this.blockRootHex,
+          mismatchedRoot: blockRootHex,
+          source: source.source,
+          peerId: `${source.peerIdStr}`,
+        },
+        "addBlock blockRootHex does not match BlockInput.blockRootHex"
+      );
+    }
+
+    for (const {blobSidecar} of this.blobsCache.values()) {
+      if (!blockAndBlobArePaired(block, blobSidecar)) {
+        this.blobsCache.delete(blobSidecar.index);
+        // TODO: (@matthewkeil) spec says to ignore invalid blobs but should we downscore the peer maybe?
+        // this.logger?.error(`Removing blobIndex=${blobSidecar.index} from BlockInput`, {}, err);
+      }
+    }
+
+    const hasAllData = this.blobsCache.size === block.message.body.blobKzgCommitments.length;
+
+    this.state = {
+      ...this.state,
+      hasAllData,
+      block,
+      versionHashes: getVersionHashes(block),
+      source,
+      timeCompleteSec: hasAllData ? source.seenTimestampSec : undefined,
+    } as BlockInputBlobsState;
+    this.blockPromise.resolve(block);
+    if (hasAllData) {
+      this.dataPromise.resolve(this.getBlobs());
+    }
+  }
+
+  hasBlob(blobIndex: BlobIndex): boolean {
+    return this.blobsCache.has(blobIndex);
+  }
+
+  addBlob({blockRootHex, blobSidecar, source, peerIdStr, seenTimestampSec}: AddBlob): void {
+    if (this.state.hasAllData) {
+      throw new BlockInputError(
+        {
+          code: BlockInputErrorCode.INVALID_CONSTRUCTION,
+          blockRoot: this.blockRootHex,
+        },
+        "Cannot addBlob to BlockInputBlobs after it already is complete"
+      );
+    }
```

Comment:
Im not sure if we should throw here but open to the idea.  I guess if we do we may want to check for the specific index is present already??  🤷‍♂️ 

```ts
if (this.blobsCache.has(blobSidecar.index)) {
```